Sources/SentrySwiftUI/SentryTracedView.swift

```diff
@@ -38,62 +40,122 @@ import SwiftUI
+#endif
         }
-
-        let id = SentryPerformanceTracker.shared.startSpan(withName: transactionCreated ? "\(self.name) - body" : self.name, nameSource: nameSource, operation: "ui.load", origin: self.traceOrigin)
-
-        SentryPerformanceTracker.shared.pushActiveSpan(id)
+        
         defer {
-            SentryPerformanceTracker.shared.popActiveSpan()
-            SentryPerformanceTracker.shared.finishSpan(id)
+            if let spanId = spanId { finishSpan(spanId) }
         }
-
-        return self.content().onAppear {
-            self.viewAppeared = true
+        
+        return content()
+            .onAppear {
+                if !viewAppeared {
+                    viewAppeared = true
+                    viewModel.tracker?.reportInitialDisplay()
+                }
+            }
+    }
+#if canImport(SwiftUI) && canImport(UIKit) && os(iOS) || os(tvOS)
+    private func startTTDTraker(for trace: SentryTracer) {
+        let tracker = SentryTimeToDisplayTracker(name: self.name, waitForFullDisplay: self.waitforFullDisplay)
+        SentryUIViewControllerPerformanceTracker.shared.setTimeToDisplay(tracker)
+        tracker.start(for: trace)
+        
+        viewModel.tracker = tracker
+    }
+#endif
+    
+    private func ensureTransactionExists() -> SentryTracer? {
+        guard SentryPerformanceTracker.shared.activeSpanId() == nil else { return nil }
+        
+        let transactionId = SentryPerformanceTracker.shared.startSpan(
+            withName: name,
+            nameSource: nameSource,
+            operation: "ui.load",
+            origin: traceOrigin
+        )
+        SentryPerformanceTracker.shared.pushActiveSpan(transactionId)
+        
+        //According to Apple's documentation, the call to body needs to be fast
+        //and can be made many times in one frame. Therefore they don't use async code to process the view.
+        //Scheduling to finish the transaction at the end of the main loop seems the least hack solution right now.
+        //'onAppear' is not a suitable place to do this because it may happen before other view body property get called.
+        DispatchQueue.main.async {
+            self.finishSpan(transactionId)
         }
+        
+        return SentryPerformanceTracker.shared.getSpan(transactionId) as? SentryTracer
+    }
+    
+    private func createAndPushBodySpan(transactionCreated: Bool) -> SpanId {
+        let spanName = transactionCreated ? "\(name) - body" : name
+        let spanId = SentryPerformanceTracker.shared.startSpan(
+            withName: spanName,
+            nameSource: nameSource,
+            operation: "ui.load",
+            origin: traceOrigin
+        )
+        SentryPerformanceTracker.shared.pushActiveSpan(spanId)
+        return spanId
+    }
+    
+    private func finishSpan(_ spanId: SpanId) {
+        SentryPerformanceTracker.shared.popActiveSpan()
+        SentryPerformanceTracker.shared.finishSpan(spanId)
```

Comment:
`m`:  There is so much logic in this code in here. I would appreciate it if we could extract the logic into an extra testable component and/or write tests for all the different types of edge cases. Currently, I don't know how I can trust this code without having a good test suite covering multiple edge cases.